src/generator/passes.jl

```diff
@@ -995,6 +997,22 @@ function (x::ProloguePrinter)(dag::ExprDAG, options::Dict)
             println(io, "using CEnum")
             println(io)
         end
+
+        # printed directedly, otherwise foo(x) = y will be shown either as `foo(x) = begin y end` or `function foo(x) y end`, 
+        # both of which needs two lines.
+        if add_to_c_type_functions
+            println(io, """
+            to_c_type(t::Type{Union{Bool, Int8, Int16, Int32, Int64, UInt8, UInt16, UInt32, UInt64, Float64, Complex{Float32}, Complex{Float64}}}) = t
+            to_c_type(t::Type{<:Vector}) = Ptr{to_c_type(eltype(t))}
+            to_c_type(::Type{String}) = Cstring
+            to_c_type(::Type{<:Ref}) = Ptr{Cvoid}
+            to_c_type(t) = t
+            to_c_type_pairs(va_list) = map(enumerate(to_c_type.(va_list))) do (ind, type)
+                :(va_list[\$ind]::\$type)
+            end
```

Comment:
Yes I now have realized that most of the conversions are just no-op, but at least Julia's String and Array type need some annotation to work: user must specify `Cstring` or `Ptr`, otherwise they are treated like normal Julia structs.

So I could make this hack optional, and document that the normal conversion of String and Array are not usually what is intended and the user is responsible for converting the argument to exactly what c code wants.

Reply:
> but at least Julia's String and Array type need some annotation to work: user must specify `Cstring` or `Ptr`, otherwise they are treated like normal Julia structs.

Yeah. Indeed. But it's not trivial to make it right. I guess it's easier to directly use the original `@ccall` rather than the generated function version in use cases like this one.

Reply:
As `varargs` in C is not ubiquitously used. Use cases like vectors will probably never be hit in practice.

Reply:
Yes, I believe the most important use case not covered by `cconvert` and `unsafe_convert` is only String, then mentioning it in the documentation suffices.